Normative: Parent completion ordering DFS invariant extension

guybedford opened the pull request:
This extension provides an approach to the case brought up in https://github.com/tc39/proposal-top-level-await/issues/158 where parent completion orderings entirely differing from the normal post-order execution might be seen as surprising behaviour.

This approach uses a new global execution ordering counter stored on the realm to index every async execution with its initial execution ordering as a global execution ordering allowing all module executions in the same realm to be sorted.

The async fulfillment is then split out into two separate phases: a first phase to gather the list of parent modules that are synchronously ready for execution at async fulfillment time of a dependent, and a second phase to execute this list in global execution-sorted order based on the new [[AsyncExecIndex]] field set during the inital DFS algorithm.

Rejections continue to be handled as previously, including the specially handled cycle rejection checks.

A global ordering numbering is necessary because multiple top-level graphs can be interacting at the async fulfillment level.

This change only affects execution timings of deferred async executions, imposing this new post-ordering sort on them. If we consider module graphs where every module uses top-level await, using the same well-defined post-order execution ordering on all subgraph execution completions that run together seems like a very sensible adjustment to me to ensure predictable user behaviours.

Further review / feedback very welcome. // cc @Ms2ger @littledan @MylesBorins @robpalme @sokra

guybedford commented on spec.html:
```diff
@@ -17,6 +17,107 @@ <h1>Introduction</h1>
+    </tr>
+    <tr>
+      <td>
+        <ins>[[AsyncModuleExecIndex]]</ins>
```
I'm not quite sure what to call this as there is no precendent for global numberings in the spec.

I did quite like the idea of defining it only to be greater than any in-flight execution, but perhaps that is too nuanced and simply treating it as a global module exec count would be more straightforward.

jorendorff replied:
Guy, consider using language like "in the order in which they were marked as blocked".

Often the spec is hard to comprehend, and easier to implement without understanding it, because it is written in such a low-level procedural style. This affects promises and modules especially.

There are opportunities to use more declarative language, to make it much clearer what the spec is trying to do, without introducing imprecision or nondeterminism.

To be clear this is just a difference in style—the specification would remain semantically the same. But I think it would have effects. More people would be able to understand the spec—in terms of time investment and in terms of literal mental capacity. I count myself in this category. Implementation techniques may end up differing more, which I consider a net positive. (In this case, for example, an implementation might be able to find a way to keep the modules in a sorted data structure, possibly eliminating the O(n log n) sort.) The only drawback for the standard's purposes is that these tables would no longer necessarily exhaustively cover every bit of state in a Realm. But I doubt that is particularly helpful anyway, once you get this far into details of particular features.

guybedford replied:
@jorendorff thanks for the feedback on this, any way we can improve or clarify are very welcome.

Are you suggesting for this case not to mark it on the realm, and instead define the execIndex on the module records as something that is just a more general monotonic comparison over all in flight module records? Or just that we change the description of this field here to capture that the ordering is the post-order of the DFS evaluation?

guybedford replied:
I've also added more notes to some of the intermediate functions explaining their roles.

jorendorff replied:
I'm proposing dropping Realm.[[AsyncModuleExecIndex]] and Module.[[AsyncExecIndex]]. I believe they are both used solely to specify the ordering in AsyncModuleExecutionFulfilled. But integers are not the only way to describe an ordering. I think the same ordering can be specified unambiguously in prose, and it would be clearer.

Instead of these steps in InnerModuleEvaluation:

> 1. If _module_.[[PendingAsyncDependencies]] &gt; 0 or _module_.[[Async]] is *true*, then
>   1. Let _execIndex_ be _module_.[[Realm]].[[AsyncModuleExecIndex]].
>   1. Set _module_.[[AsyncExecIndex]] to _execIndex_.
>   1. Set _module_.[[Realm]].[[AsyncModuleExecIndex]] to _execIndex_ + 1.
>   1. Set _module_.[[AsyncEvaluating]] to *true*.

I think it could say:

> 1. If _module_.[[PendingAsyncDependencies]] &gt; 0 or _module_.[[Async]] is *true*, then
>   1. Assert: _module_.[[AsyncEvaluating]] is *false* and has never been *true*.
>   1. Set _module_.[[AsyncEvaluating]] to *true*. NOTE: The module becomes blocked. The order in which modules become blocked is significant; see AsyncModuleExecutionFulfilled.

And in AsyncModuleExecutionFulfilled, instead of:

> 1. Let _sortedExecList_ be a List of elements that are the elements of _execList_ ordered by the [[AsyncModuleExecIndex]] field of the Cyclic Module Records in ascending numerical order.

it could say:

> 1. Assert: each element of _execList_ is a Cyclic Module Record with the [[AsyncEvaluating]] field set to *true*.
> 1. Let _sortedExecList_ be a List of the elements of _execList_ , in the order in which their [[AsyncEvaluating]] fields were set to *true*.

If this doesn't seem obviously better, no need to bother with it! But to me this is clearer, and the information is spread across two places in the spec rather than four.

guybedford replied:
This is much nicer yes, I wouldn't have thought that would be have been a well-defined enough definition, but it is certainly a precise one and if it works then certainly let's do this. Thanks for the suggestion!

Ms2ger commented:
Yulia: I think you've got more of the proposal paged in than me right now, so I'd appreciate it if you could take a look. Let me know if you prefer I try.

codehag commented on spec.html:
```diff
@@ -686,23 +812,40 @@ <h1><ins>AsyncModuleExecutionFulfilled ( _module_ )</ins></h1>
+        1. Let _cycleRoot_ be ! GetAsyncCycleRoot(_m_).
+        1. Assert: _cycleRoot_.[[EvaluationError]] is *undefined*.
+        1. Let _cycleModules_ be a new empty List.
+        1. Perform ! GatherCycleModules(_cycleRoot_, _cycleRoot_.[[DFSIndex]], _cycleModules_).
```
It looks like this does solve the, I guess we can say fragility? That I came across in my issue. I want to work with this a bit more and make sure that I understand it. Will do that tomorrow.

guybedford replied:
The root cause of the fragility here was most likely due to https://github.com/tc39/proposal-top-level-await/pull/161.

guybedford commented:
I've rebased this to the PR bug fix in https://github.com/tc39/proposal-top-level-await/pull/159, and also separated out the cycle gathering on failure, as cycle errors are allowed to be sparse in the current TLA algorithm, it doesn't seem such an issue to ensure consistency by just relying on cycle root. Instead we can follow this up in a separate PR if necessary.

Thanks for feedback so far. The diff should now be a little simpler as well.

jorendorff commented on spec.html:
```diff
@@ -639,54 +768,52 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+      <emu-alg>
+        1. Let _execList_ be a new empty List.
+        1. Perform ! GatherAsyncParentCompletions(_module_, _execList_).
+        1. Let _sortedExecList_ be a List of elements that are the elements of _execList_ ordered by the [[AsyncModuleExecIndex]] field of the Cyclic Module Records in ascending numerical order.
```
Typo - the field is named [[AsyncExecIndex]].

guybedford commented:
@jorendorff I've included your suggestions - brings the diff down nicely as well.

guybedford commented:
I've rebased this PR to the latest merge of the cycle root fix, should be much simpler to review. Further feedback welcome, I do think this is an important normative change to make. If it needs to go explicitly through the meeting that would be understandable.

Further feedback would be very welcome to understand implementer interest.

codehag reviewed (commented):
I have unfortunately run out of time this week to fully do this review. If it can wait a week, I will do it then -- or someone can also take over. I made a few small comments.

I want to still make sure that we don't inadvertently end up changing this behaviour: https://github.com/web-platform-tests/wpt/pull/26771/files

Guy says it shouldn't be an issue but I would like to take more time to verify the implications.

codehag commented on spec.html:
```diff
@@ -592,8 +593,11 @@ <h1>InnerModuleEvaluation ( _module_, _stack_, _index_ )</h1>
         1. <del>Perform ? _module_.ExecuteModule().</del>
-        1. <ins>If _module_.[[PendingAsyncDependencies]] &gt; 0, set _module_.[[AsyncEvaluating]] to *true*.</ins>
-        1. <ins>Otherwise, if _module_.[[Async]] is *true*, perform ! ExecuteAsyncModule(_module_).</ins>
+        1. <ins>If _module_.[[PendingAsyncDependencies]] &gt; 0 or _module_.[[Async]] is *true*, then</ins>
```
We will end up wrapping non-async modules in promises, due to this line. Not sure yet how I feel about it.

guybedford replied:
This is mostly an editorial refactoring. `[[AsyncEvaluating]]` was formerly set in `ExecuteAsyncModule`, we bring that set out of that function and set it here for both of the previous two conditional paths.

I only wanted to combine these two conditions into one because in both cases we want the same ordering note to apply, which otherwise would have needed to be repeated in both paths.

So the only semantic change really is that note, the rest is just editorial refactoring to permit a single note.

guybedford replied:
Strictly speaking, the order of [[AsyncEvaluating]] does only matter in the [[PendingAsyncDependencies]] > 0 case so that we could just have the note in the first condition and leave the logic as before with the diff just being a note on the first line of this diff.

codehag replied:
OK, good! I most definitely was surprised because it didn't fit my mental model.

guybedford commented on spec.html:
```diff
@@ -592,8 +593,11 @@ <h1>InnerModuleEvaluation ( _module_, _stack_, _index_ )</h1>
-        1. <ins>If _module_.[[PendingAsyncDependencies]] &gt; 0, set _module_.[[AsyncEvaluating]] to *true*.</ins>
-        1. <ins>Otherwise, if _module_.[[Async]] is *true*, perform ! ExecuteAsyncModule(_module_).</ins>
+        1. <ins>If _module_.[[PendingAsyncDependencies]] &gt; 0 or _module_.[[Async]] is *true*, then</ins>
+          1. <ins>Assert: _module_.[[AsyncEvaluating]] is *false* and was never previously set to *true*.</ins>
```
This was an additional assertion suggested by @jorendorff here, I'm open to rewording that could make it seem more clear.

codehag commented on spec.html:
```diff
@@ -652,28 +655,44 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
           1. Assert: _module_.[[EvaluationError]] is not *undefined*.
           1. Return *undefined*.
-        1. Assert: _module_.[[EvaluationError]] is *undefined*.
-        1. Set _module_.[[AsyncEvaluating]] to *false*.
```
Are we no longer set asyncEvaluating to false? In this case, maybe the flag name should be changed... like "PendingAsyncEvaluation" or something...

guybedford replied:
This was an oversight, thanks, I've added it in a follow-up iteration now. I do like the idea of `PendingAsyncEvaluation` though and that might be a possible future spec simplification if it can handle the required state checks. The `AsyncEvaluating` definition as currently is supposed to be as defined though.

codehag reviewed (commented):
Still thinking about the order of evaluation. A couple of comments so far

guybedford commented on spec.html:
```diff
@@ -652,28 +655,44 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+        1. Let _sortedExecList_ be a List of elements that are the elements of _execList_, in the order in which they had their [[AsyncEvaluating]] fields set to *true* in InnerModuleEvaluation.
+        1. Assert: All elements of _sortedExecList_ have their [[AsyncEvaluating]] field set to *true*, [[PendingAsyncDependencies]] field set to 0 and [[EvaluationError]] field set to *undefined*.
+        1. For each Module _m_ of _sortedExecList_, do
+          1. If _m_.[[AsyncEvaluating]] is *false*, then
```
This is to handle cascading errors up the tree that may have been in the execution list to ensure we don't evaluate an errored module.

guybedford replied:
(The state change would have been initiated by the error fulfillment call in this loop)

codehag commented on spec.html:
```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+            1. Let _result_ be _m_.ExecuteModule().
+            1. If _result_ is an abrupt completion,
+              1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+        1. For each Module _m_ of _sortedExecList_, do
```
Maybe we can bundle this with the above loop?

guybedford replied:
Yes, and also we're only supposed to transition the state to false for the sync executions as well so I've added that check too.

codehag commented on spec.html:
```diff
@@ -648,28 +651,45 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
         1. For each Module _m_ of _module_.[[AsyncParentModules]], do
-          1. Decrement _m_.[[PendingAsyncDependencies]] by 1.
-          1. If _m_.[[PendingAsyncDependencies]] is 0 and _m_.[[EvaluationError]] is ~empty~, then
+          1. If _execList_ does not contain _m_ and _m_.[[CycleRoot]].[[EvaluationError]] is ~empty~, then
```
I think I have confirmed this now: we probably need to also guard here on either "AsyncEvaluating" being true, or "PendingAsyncDependencies" is greater than 0. In my test implementation I went with the later, otherwise we may run this again on a module, putting pendingAsyncDependencies into an invalid state (-1).

guybedford replied:
Any module in AsyncParentModules cannot be a successful execution if we know the child is not successfully executed, and this invariant should prevent the case you describe?

If we called this function on an arbitrary module that was in an arbitrary state then I can appreciate that, but if the module is definitely a module that has not completed yet, then it's parent completions certainly won't have completed yet. The only case left should be if there was an error, where we are specifically just checking and handling that case already here.

guybedford replied:
Would it help to add a `Assert: _module_.[[AsyncEvaluating]] is equal to *true*` before we first call this function on the top _module_ in AsyncModuleExecutionFulfilled?

codehag replied:
Ah, yes it turns out it was my mistake. Probably the assertion would help

guybedford replied:
Ok great, added.

codehag commented on spec.html:
```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+            1. Let _result_ be _m_.ExecuteModule().
+            1. If _result_ is an abrupt completion,
+              1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+            1. Otherwise, set _m_.[[AsyncEvaluating]] to *false*.
```
Ok, the bug I mentioned yesterday actually has it's root here.

When we set module to false here, we are doing it before it calls it's own GatherAsyncParentCompletions, and we end up checking for an error which doesnt exist. That throws on the assertion.

We should instead move this flag change to after this loop, and only set it for the current module, because all modules (async or no) will now call asyncModuleExecutionFulfilled.

guybedford replied:
AsyncModuleExecutionFulfilled should only be called for modules that have `[[Async]]` set to true, as we only run `ExecAsyncModule` on these async modules.

The `AsyncModuleExecutionFulfilled` always can only happen after the promise completion of the module body, so we are guaranteed that this routine should finish synchronously before any of the completions trigger.

It is only for the sycn modules that do not go through async exec that we thus set `AsyncEvaluating` to false explicitly here.

Let me know if that clarifies the logic?

codehag replied:
Oh, i think this changed since I last looked at it. Let me look again

codehag replied:
Yes, this is fixed in this version! I had the old behavior from before 279739dc10d43acb86c471632a780c3c1d7a1f5f

guybedford replied:
Ah yes, that was actually fixed with your previous feedback in https://github.com/tc39/proposal-top-level-await/pull/159#discussion_r576748351.

codehag reviewed (approved):
Bar the issue I found with three tests, that appears to be a js shell issue (will verify that today), and the bug with eagerly setting asyncEvaluated on children before they finish their evaluation, this looks fine and is probably a worthwhile change. It doesn't appear to have web platform implications so we are good there. I am setting this to approved, after fixing the early setting of the flag. 

Sketch implementation here: https://phabricator.services.mozilla.com/D105437

guybedford commented:
@syg it would be great to get your review on this PR if you're able to take a look.

codehag commented on spec.html:
```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+            1. Let _result_ be _m_.ExecuteModule().
+            1. If _result_ is an abrupt completion,
+              1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+            1. Otherwise, set _m_.[[AsyncEvaluating]] to *false*.
```
Hm, this still feels wrong though.

Won't this mean that a given leaf node will not update asyncEvaluating to false, and always be set to true, unless something depends on it?

guybedford replied:
Yes, we should be handling the immediate fulfillment. Added https://github.com/tc39/proposal-top-level-await/pull/159/commits/b51c04870ed0a88f923763bcb6f6005aab43e8fe.

codehag commented:
(sorry about this) Ok, after thinking a bit more about this, what we are doing here is reverting a decision that was made earlier.

We are going from variant b ([our current semantics](https://github.com/tc39/proposal-top-level-await#semantics-as-desugaring))

```js
(async () => {
  const [a, b, c] = await Promise.all([
    import('./a.mjs'),
    import('./b.mjs'),
    import('./c.mjs')
  ]);
  console.log(a, b, c);
})();
```

back to variant a:

```js
(async () => {
  const a = await import('./a.mjs');
  const b = await import('./b.mjs');
  const c = await import('./c.mjs');
  
  console.log(a, b, c);
})();
```

Part of the motivation of this change is that this was an oversight, but I don't think so. We decided against this originally. The discussion is here: [notes](https://github.com/tc39/notes/blob/master/meetings/2019-03/mar-28.md#top-level-await-with-a-vengeance)[Slides](https://docs.google.com/presentation/d/1Jz86ztxFnNVwch50GEjmTPt1IrA0ZVBqkpzMsiSLF20/edit#slide=id.g5438938e3a_0_92)

I don't know if it makes sense to rehash this -- Myles made a strong case for this synchronous behaviour based on the usecase.

guybedford commented:
Just to reiterate, this PR is only to ensure that all sync subgraphs that depend on async modules execute in the original ordering, and it doesn't affect any of the other execution properties of this proposal that have already been agreed.

Ms2ger commented on spec.html:
```diff
@@ -648,28 +651,45 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+          1. If _execList_ does not contain _m_ and _m_.[[CycleRoot]].[[EvaluationError]] is ~empty~, then
+            1. Assert: _m_.[[EvaluationError]] is ~empty~.
+            1. Assert: _m_.[[AsyncEvaluating]] is *true*.
+            1. Set _m_.[[PendingAsyncDependencies]] to _m_.[[PendingAsyncDependencies]] - 1.
```
```suggestion
            1. Assert: 0 < _m_.[[PendingAsyncDependencies]].
            1. Set _m_.[[PendingAsyncDependencies]] to _m_.[[PendingAsyncDependencies]] - 1.
```

ljharb commented on spec.html:
```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+          1. If _execList_ does not contain _m_ and _m_.[[CycleRoot]].[[EvaluationError]] is ~empty~, then
+            1. Assert: _m_.[[EvaluationError]] is ~empty~.
+            1. Assert: _m_.[[AsyncEvaluating]] is *true*.
+            1. Assert: 0 < _m_.[[PendingAsyncDependencies]].
```
let’s avoid a yoda condition :-)
```suggestion
            1. Assert: _m_.[[PendingAsyncDependencies]] > 0.
```

syg commented:
I started implementing this an ran into a concern. I'd like the postorder position to be scoped to each top-level call to `Evaluate()`. However, this doesn't seem possible because the same async module can participate in concurrent, multiple top-level calls to `Evaluate()`, e.g. multiple `<script>` tags. That presents a problem: the postorder would need to be maintained as a global serial number or something, which puts it at higher risk of overflowing. There can be cleverer designs, I suppose, of resetting the global serial to `0` after all outstanding modules are finished, but that's too much complexity for my liking.

syg commented on spec.html:
```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+        1. Let _execList_ be a new empty List.
+        1. Perform ! GatherAsyncParentCompletions(_module_, _execList_).
+        1. Let _sortedExecList_ be a List of elements that are the elements of _execList_, in the order in which they had their [[AsyncEvaluating]] fields set to *true* in InnerModuleEvaluation.
+        1. Assert: All elements of _sortedExecList_ have their [[AsyncEvaluating]] field set to *true*, [[PendingAsyncDependencies]] field set to 0 and [[EvaluationError]] field set to *undefined*.
```
There's a bug here: the top-level capability might never get resolved. First part of the fix is to insert a step here that does:

`1. Let _topLevelParent_ be _module_.`

syg commented on spec.html:
```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+            1. If _result_ is an abrupt completion,
+              1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+            1. Otherwise, set _m_.[[AsyncEvaluating]] to *false*.
         1. If _module_.[[TopLevelCapability]] is not ~empty~, then
```
Continuing here, all references to _module_ from this step and below should be replaced with _topLevelParent_.

If an async module didn't get any parents from GatherAsyncParentCompletions, then it itself must be an entry point so resolve its top-level capability. If it did get parents, since it was async, the last parent in _execList_ might be the entry point, so check if it is and resolve the top-level capability.

syg commented on spec.html:
```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+            1. Let _result_ be _m_.ExecuteModule().
+            1. If _result_ is an abrupt completion,
+              1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+            1. Otherwise, set _m_.[[AsyncEvaluating]] to *false*.
```
Oops, forgot a new step here for the fix:

`1. Let _topLevelParent_ be _m_.`

guybedford commented:
@syg nicely spotted - yes with the flattening of the previous recursive algorithm we need to now explicitly call the capabilities on the sync module parents of async modules.

Further to your suggestion, the fix can possibly actually resolve capabilities on all sync parents of async modules, as they could all have eg a later dynamic import expression that attached to those modules specifically. For this reason I've added the capability resolution within the main exec loop to all sync modules.

I also went ahead and moved the main "top level parent" capability resolution to the top, this is to ensure that the current module promise will resolve _before_ the sync parent promises. Since all capability resolutions, I believe, wiill go into the task queue, the resolution functions can't be called synchronously. So having the capability resolution at the top should not break the sync invariant between the parent executions by calling a resolve function earlier than the parent executions. Please verify that I'm correct in that assumption though, and just let me know if further clarification would help.

Will also await on @codehag when she gets back for review before merging so we have time here too.

syg commented:
> Further to your suggestion, the fix can possibly actually resolve capabilities on all sync parents of async modules, as they could all have eg a later dynamic import expression that attached to those modules specifically. 

Great point!

syg commented:
If someone wants to an executable implementation in V8 to check things with, here's the CL: https://chromium-review.googlesource.com/c/v8/v8/+/2757901

sokra commented:
If somebody knows how to build a node.js version from that, you could run https://github.com/evanw/tla-fuzzer with that change and check the result. That's how I found the original case.

codehag reviewed (approved):
Looks good to me.

guybedford commented:
I tested the patch in Node.js and can confirm it fixes https://github.com/tc39/proposal-top-level-await/issues/158 and has the correct timings for dynamic import registrations with the latest commit as well.

jugglinmike commented:
We've merged a test for the new semantics to Test262: https://github.com/tc39/test262/pull/2989
